woocommerce/woocommerce
# Product Editor Dev Tools: Improved expression evaluation tooling

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR enhances the Product Editor Dev Tools that are part of the WooCommerce Beta Tester with the following new functionality:

- Ability to edit existing expressions added in the tool
- Ability to remove expressions added in the tool
- Show expression evaluation result live as the expression is typed
- Show the expression evaluation result type
- Improve the display of expression evaluation result types like object

_Note: Because this dev tool is part of the WooCommerce Beta Tester, which is [not officially supported](https://developer.woocommerce.com/2023/08/23/woocommerce-beta-tester-plugin-deprecation-vulnerability-found/), the same level of rigor with UX design has not been taken. There are some rough edges that should be improved at some point, but I believe these changes, as they are implemented, are a huge improvement over the existing dev tool, and very useful for developers working with the new product editor._

Closes #48544.

#### Before

https://github.com/woocommerce/woocommerce/assets/2098816/a317d19a-5594-4eaa-8987-7880818818e2

#### After

https://github.com/woocommerce/woocommerce/assets/2098816/b5091590-fda5-4324-b7bf-63aea813311e



### How to test the changes in this Pull Request:



Note: You need to build the `WooCommerce Beta Tester` plugin ([README](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-beta-tester/README.md)) with this branch and run it to get these updates.

With a WooCommerce env with the new product editor enabled (**WooCommerce** > **Settings** > **Advanced** > **Features** > **Experimental features**)...

1. Go to **Products** > **Add New**
2. Click on the **Show developer tools** menu item in the more menu (the kebab/ellipsis menu)
3. Go to the **Product** tab in the developer tools in the footer
4. Try adding some expressions, editing then, removing them and verify things work as expected.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
